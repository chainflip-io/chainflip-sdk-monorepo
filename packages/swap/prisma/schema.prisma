// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  uuid                  String   @unique @default(uuid())
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal              @db.Decimal(30, 0)
  intermediateAmount        Decimal?             @db.Decimal(30, 0)
  egressAmount              Decimal?             @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  egressCompletedAt         DateTime?
  egressCompletedBlockIndex String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?             @unique
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  chain       Chain
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  chain    Chain
  egresses Egress[]

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
