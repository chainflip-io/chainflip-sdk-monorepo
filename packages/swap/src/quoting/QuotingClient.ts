import * as crypto from 'crypto';
import { EventEmitter } from 'events';
import { io, Socket } from 'socket.io-client';
import { promisify } from 'util';
import { AssetAndChain } from '@/shared/enums';
import { LegJson, MarketMakerQuoteRequest, MarketMakerRawQuote } from './schemas';
import logger from '../utils/logger';

const signAsync = promisify(crypto.sign);

export type QuoteHandler = (
  quote: MarketMakerQuoteRequest<LegJson>,
) => Promise<MarketMakerRawQuote>;

/**
 * A reference implementation of a client that connects to the quoting service
 * and handles quote requests
 */
export default class QuotingClient extends EventEmitter {
  private socket!: Socket;

  private quoteHandler!: QuoteHandler;

  private privateKey: crypto.KeyObject;

  constructor(
    private readonly url: string,
    private readonly accountId: string,
    privateKey: string,
    private readonly quotedAssets?: AssetAndChain[],
  ) {
    super();
    this.privateKey = crypto.createPrivateKey({
      key: Buffer.from(privateKey),
      format: 'pem',
      type: 'pkcs8',
    });
  }

  private async getAuth() {
    const timestamp = Date.now();
    return {
      timestamp,
      client_version: this.quotedAssets !== undefined ? '2' : '1',
      account_id: this.accountId,
      signature: await this.getSignature(timestamp),
      quoted_assets: this.quotedAssets,
    };
  }

  async connect() {
    this.socket = io(this.url, {
      transports: ['websocket'],
      auth: await this.getAuth(),
    });

    this.socket.on('connect', () => {
      if (this.socket.connected) {
        logger.info('connected to quoting service');
      }
      this.emit('connected');
    });

    this.socket.on('quote_request', async (quote: MarketMakerQuoteRequest<LegJson>) => {
      const response = await this.quoteHandler(quote);
      this.socket.emit('quote_response', {
        ...response,
        request_id: quote.request_id,
      } as MarketMakerRawQuote);
    });

    this.socket.on('connect_error', async (err) => {
      // the reason of the error, for example "xhr poll error"
      logger.error('connect_error', err.message);
      // retry with a new signature
      const updatedTimestamp = Date.now();
      this.socket.auth = {
        timestamp: updatedTimestamp,
        client_version: '1',
        account_id: this.accountId,
        signature: await this.getSignature(updatedTimestamp),
      };
    });

    this.socket.on('disconnect', () => {
      this.emit('disconnect');
    });
  }

  private async getSignature(timestamp: number): Promise<string> {
    const buffer = await signAsync(
      null,
      Buffer.from(`${this.accountId}${timestamp}`, 'utf8'),
      this.privateKey,
    );

    return buffer.toString('base64');
  }

  setQuoteRequestHandler(handler: QuoteHandler) {
    this.quoteHandler = handler;
  }

  close() {
    this.socket.close();
  }
}
