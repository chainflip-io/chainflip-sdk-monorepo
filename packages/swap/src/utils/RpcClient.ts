import assert from 'assert';
import EventEmitter, { once } from 'events';
import { filter, firstValueFrom, Subject, timeout } from 'rxjs';
import WebSocket from 'ws';
import { z } from 'zod';
import { SupportedAsset } from '@/shared/enums';
import { handleExit, waitWithTimeout } from './function';
import logger from './logger';

const READY = 'READY';
const DISCONNECT = 'DISCONNECT';

export type RpcAsset = {
  [K in SupportedAsset]: Capitalize<Lowercase<K>>;
}[SupportedAsset];

type RpcResponse =
  | { id: number; result: unknown }
  | { id: number; error: { code: number; message: string } };

export default class RpcClient<
  Req extends Record<string, z.ZodTypeAny>,
  Res extends Record<string, z.ZodTypeAny>,
> extends EventEmitter {
  private socket!: WebSocket;

  private requestId = 0;

  private messages = new Subject<RpcResponse>();

  private connectionFailures = 0;

  constructor(
    private readonly url: string,
    private readonly requestMap: Req,
    private readonly responseMap: Res,
    private readonly namespace: string,
  ) {
    super();
    handleExit(() => this.handleClose());
  }

  async close() {
    await this.handleClose();
  }

  private async handleClose() {
    this.socket.removeAllListeners('close');
    this.socket.close();
    if (this.socket.readyState !== WebSocket.CLOSED) {
      await once(this.socket, 'close');
    }
  }

  private async connectionReady() {
    if (this.socket.readyState === WebSocket.OPEN) return;
    await waitWithTimeout(once(this, READY), 30000);
  }

  async connect(): Promise<this> {
    this.socket = new WebSocket(this.url);
    this.socket.on('message', (data) => {
      this.messages.next(JSON.parse(data.toString()));
    });

    // this event is also emitted if a socket fails to open, so all reconnection
    // logic will be funnelled through here
    this.socket.once('close', async () => {
      this.emit(DISCONNECT);

      const backoff = Math.min(250 * 2 ** this.connectionFailures, 30000);

      logger.info(`websocket closed, reconnecting in ${backoff}ms`);

      setTimeout(() => {
        this.connect().catch(() => {
          this.connectionFailures += 1;
        });
      }, backoff);
    });

    this.socket.on('error', (error) => {
      logger.customError('received websocket error', {}, { error });
      this.socket.close();
    });

    if (this.socket.readyState !== WebSocket.OPEN) {
      await waitWithTimeout(once(this.socket, 'open'), 30000);
    }

    this.emit(READY);
    this.connectionFailures = 0;

    return this;
  }

  async sendRequest<R extends keyof Req & keyof Res>(
    method: R,
    ...params: z.input<Req[R]>
  ): Promise<z.infer<Res[R]>> {
    let response: RpcResponse | undefined;

    for (let i = 0; i < 5; i += 1) {
      try {
        const id = this.requestId;
        this.requestId += 1;

        await this.connectionReady();

        this.socket.send(
          JSON.stringify({
            id,
            jsonrpc: '2.0',
            method: `${this.namespace}_${method as string}`,
            params: this.requestMap[method].parse(params),
          }),
        );

        response = await Promise.race([
          firstValueFrom(
            this.messages.pipe(
              filter((msg) => msg.id === id),
              timeout(30000),
            ),
          ),
          // if the socket closes after sending a request but before getting a
          // response, we need to retry the request
          once(this, DISCONNECT).then(() => {
            throw new Error('disconnected');
          }),
        ]);

        break;
      } catch {
        // retry
      }
    }

    assert(response, 'no response received');

    if ('error' in response) throw new Error(response.error.message);

    return this.responseMap[method].parse(response.result) as z.infer<Res[R]>;
  }
}
