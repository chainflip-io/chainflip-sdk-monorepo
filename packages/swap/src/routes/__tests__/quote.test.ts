import axios from 'axios';
import * as crypto from 'crypto';
import { once } from 'events';
import { Server } from 'http';
import { AddressInfo } from 'net';
import request from 'supertest';
import { promisify } from 'util';
import RpcClient from '@/shared/node-apis/RpcClient';
import { swappingEnvironment } from '@/shared/tests/fixtures';
import prisma from '../../client';
import QuotingClient from '../../quoting/QuotingClient';
import app from '../../server';

const generateKeyPairAsync = promisify(crypto.generateKeyPair);

jest.mock(
  '@/shared/node-apis/RpcClient',
  () =>
    class {
      async connect() {
        return this;
      }

      sendRequest() {
        throw new Error('unmocked request');
      }
    },
);

jest.mock('@/shared/consts', () => ({
  ...jest.requireActual('@/shared/consts'),
  getMinimumSwapAmount: jest.fn().mockReturnValue('100'),
  getPoolsNetworkFeeHundredthPips: jest.fn().mockReturnValue(1000),
}));

jest.mock('axios', () => ({
  post: jest.fn(() =>
    Promise.resolve({
      data: swappingEnvironment(),
    }),
  ),
}));
describe('server', () => {
  let server: Server;
  let client: QuotingClient;

  beforeAll(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE public."Pool" CASCADE`;
    await prisma.pool.createMany({
      data: [
        {
          baseAsset: 'USDC',
          pairAsset: 'FLIP',
          liquidityFeeHundredthPips: 1000,
        },
        {
          baseAsset: 'USDC',
          pairAsset: 'ETH',
          liquidityFeeHundredthPips: 2000,
        },
      ],
    });
  });

  beforeEach(async () => {
    server = app.listen(0);
    await prisma.$queryRaw`TRUNCATE TABLE private."MarketMaker" CASCADE`;
    const name = 'web_team_whales';
    const pair = await generateKeyPairAsync('ed25519');
    await prisma.marketMaker.create({
      data: {
        name: 'web_team_whales',
        publicKey: pair.publicKey
          .export({ format: 'pem', type: 'spki' })
          .toString('base64'),
      },
    });

    client = new QuotingClient(
      `http://localhost:${(server.address() as AddressInfo).port}`,
      name,
      pair.privateKey.export({ format: 'pem', type: 'pkcs8' }).toString(),
    );
    await once(client, 'connected');
  });

  afterEach((cb) => {
    client.close();
    server.close(cb);
  });

  describe('GET /quote', () => {
    it('rejects if amount is lower than minimum swap amount', async () => {
      jest
        .mocked(axios.post)
        .mockResolvedValueOnce({ data: swappingEnvironment('0xffffff') });

      const params = new URLSearchParams({
        srcAsset: 'FLIP',
        destAsset: 'ETH',
        amount: '50',
      });

      const { body, status } = await request(server).get(
        `/quote?${params.toString()}`,
      );

      expect(status).toBe(400);
      expect(body).toMatchObject({
        message: 'expected amount is below minimum swap amount (16777215)',
      });
    });

    it('rejects if amount is higher than maximum swap amount', async () => {
      jest
        .mocked(axios.post)
        .mockResolvedValueOnce({ data: swappingEnvironment('0x1') });

      const params = new URLSearchParams({
        srcAsset: 'USDC',
        destAsset: 'FLIP',
        amount: '50',
      });

      const { body, status } = await request(server).get(
        `/quote?${params.toString()}`,
      );

      expect(status).toBe(400);
      expect(body).toMatchObject({
        message: 'expected amount is above maximum swap amount (2)',
      });
    });

    it('gets the quote from usdc when the broker is best', async () => {
      const sendSpy = jest
        .spyOn(RpcClient.prototype, 'sendRequest')
        .mockResolvedValueOnce({
          egressAmount: (1e18).toString(),
        });

      const params = new URLSearchParams({
        srcAsset: 'USDC',
        destAsset: 'ETH',
        amount: (100e6).toString(),
      });

      client.setQuoteRequestHandler(async (req) => ({
        id: req.id,
        egress_amount: (0.5e18).toString(),
      }));

      const { body, status } = await request(server).get(
        `/quote?${params.toString()}`,
      );

      expect(status).toBe(200);
      expect(body).toMatchObject({
        id: expect.any(String),
        egressAmount: (1e18).toString(),
        includedFees: [
          { amount: '100000', asset: 'USDC', type: 'network' },
          { amount: '200000', asset: 'USDC', type: 'liquidity' },
        ],
      });
      expect(sendSpy).toHaveBeenCalledTimes(1);
    });

    it('gets the quote to usdc when the broker is best', async () => {
      const sendSpy = jest
        .spyOn(RpcClient.prototype, 'sendRequest')
        .mockResolvedValueOnce({
          egressAmount: (100e6).toString(),
        });

      const params = new URLSearchParams({
        srcAsset: 'ETH',
        destAsset: 'USDC',
        amount: (1e18).toString(),
      });

      client.setQuoteRequestHandler(async (req) => ({
        id: req.id,
        egress_amount: (50e6).toString(),
      }));

      const { body, status } = await request(server).get(
        `/quote?${params.toString()}`,
      );

      expect(status).toBe(200);
      expect(body).toMatchObject({
        id: expect.any(String),
        egressAmount: (100e6).toString(),
        includedFees: [
          { amount: '100100', asset: 'USDC', type: 'network' },
          { amount: '2000000000000000', asset: 'ETH', type: 'liquidity' },
        ],
      });
      expect(sendSpy).toHaveBeenCalledTimes(1);
    });

    it('gets the quote with intermediate amount when the broker is best', async () => {
      const sendSpy = jest
        .spyOn(RpcClient.prototype, 'sendRequest')
        .mockResolvedValueOnce({
          intermediateAmount: (2000e6).toString(),
          egressAmount: (1e18).toString(),
        });

      const params = new URLSearchParams({
        srcAsset: 'FLIP',
        destAsset: 'ETH',
        amount: (1e18).toString(),
      });

      client.setQuoteRequestHandler(async (req) => ({
        id: req.id,
        intermediate_amount: (1000e6).toString(),
        egress_amount: (0.5e18).toString(),
      }));

      const { body, status } = await request(server).get(
        `/quote?${params.toString()}`,
      );

      expect(status).toBe(200);
      expect(body).toMatchObject({
        id: expect.any(String),
        intermediateAmount: (2000e6).toString(),
        egressAmount: (1e18).toString(),
        includedFees: [
          { amount: '2000000', asset: 'USDC', type: 'network' },
          { amount: '1000000000000000', asset: 'FLIP', type: 'liquidity' },
          { amount: '4000000', asset: 'USDC', type: 'liquidity' },
        ],
      });
      expect(sendSpy).toHaveBeenCalledTimes(1);
    });

    it('gets the quote when the market maker is best', async () => {
      const sendSpy = jest
        .spyOn(RpcClient.prototype, 'sendRequest')
        .mockResolvedValueOnce({
          intermediateAmount: (2000e6).toString(),
          egressAmount: (1e18).toString(),
        });
      const params = new URLSearchParams({
        srcAsset: 'FLIP',
        destAsset: 'ETH',
        amount: (1e18).toString(),
      });

      client.setQuoteRequestHandler(async (req) => ({
        id: req.id,
        intermediate_amount: (3000e6).toString(),
        egress_amount: (2e18).toString(),
      }));

      const { body, status } = await request(server).get(
        `/quote?${params.toString()}`,
      );

      expect(status).toBe(200);
      expect(body).toMatchObject({
        id: expect.any(String),
        intermediateAmount: (2994e6).toString(),
        egressAmount: (1.992e18).toString(),
        includedFees: [
          { amount: '2994000', asset: 'USDC', type: 'network' },
          { amount: '1000000000000000', asset: 'FLIP', type: 'liquidity' },
          { amount: '5988000', asset: 'USDC', type: 'liquidity' },
        ],
      });
      expect(sendSpy).toHaveBeenCalledTimes(1);
    });
  });
});
