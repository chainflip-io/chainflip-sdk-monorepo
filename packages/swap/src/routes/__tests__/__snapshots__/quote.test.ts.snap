// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server gets the quote for deprecated params without the chain 1`] = `
{
  "egressAmount": "999999999999975000",
  "includedFees": [
    {
      "amount": "1000000",
      "asset": "FLIP",
      "chain": "Ethereum",
      "type": "INGRESS",
    },
    {
      "amount": "2000000",
      "asset": "USDC",
      "chain": "Ethereum",
      "type": "NETWORK",
    },
    {
      "amount": "999999999999000",
      "asset": "FLIP",
      "chain": "Ethereum",
      "type": "LIQUIDITY",
    },
    {
      "amount": "4000000",
      "asset": "USDC",
      "chain": "Ethereum",
      "type": "LIQUIDITY",
    },
    {
      "amount": "25000",
      "asset": "ETH",
      "chain": "Ethereum",
      "type": "EGRESS",
    },
  ],
  "intermediateAmount": "2000000000",
}
`;
